### PR TITLE
Making GlobalXPooling1D work with masked inputs

### PR DESCRIPTION
**Pull request by RaffEdwardBAH**
For some of our work we have found it useful to average over masked sequences, such as when using an LSTM over variable length data. Hoping to add the code back. Also modified it to support masking for the MaxPooling layer. 

I'm not the best at thinking in Theano ways, so I apologize if its not the best use of the code. I had some issues with the code not working with Tensorflow that was fixed by the extra casts. 

Modifying GlobalAveragePooling1D to work with maksed inputs by averaging only across the non-masked time-steps. Max handled by subtracting the largest range + overhead from masked positions.

### CODE REVIEW

**Comment by fchollet** [comment]
Failing the PEP8 test.

**Comment by RaffEdwardBAH** [comment]
Ran autopep8 on the file I altered.

File: keras/layers/pooling.py [hunk 1]
```diff
@@ -424,7 +444,21 @@ class GlobalAveragePooling1D(_GlobalPooling1D):
...
     def call(self, x, mask=None):
-        return K.mean(x, axis=1)
+        if mask is None:
+            return K.mean(x, axis=1)
+        else:
+            mask = K.cast(mask, K.floatx())
+            # Expand count so we can do a broadcast division
+            count = K.sum(mask, axis=1)
+            count = K.expand_dims(count, dim=1)
+            if K.backend() == 'tensorflow':  # tensorflow shape command dosn't work
```
Comment (fchollet) [inline keras/layers/pooling.py]: We should not have this type of backend-specific branching. Why doesn't the same line work in TF? It should work, so if it doesn't then that's the problem we should solve.

Comment (RaffEdwardBAH) [inline keras/layers/pooling.py]: I don't know why just shape didn't work with tensorflow, I didn't get a helpful error message back when I wrote this code. I also had to cast the mask to a floatx because tensorflow would throw errors on it otherwise. There have been a number of threads discussions issue relating to backend code that would work with theano and not tensorflow, so I don't think this is a new issue. 

We also shouldn't have stuff like "int_shape" which is documented as part of the backend but only defined for the tensorflow backend. But it exists.

Comment (fchollet) [inline keras/layers/pooling.py]: `int_shape` only exists for TF because only TF implements it. Future backends will hopefully implement it, and maybe Theano will implement it too in the future.

Another reason why this implementation won't work: these layers are meant to work for variable-size inputs (that's basically the motivation for having them vs. using regular pooling layers). So you cannot access `int_shape`, it won't work with None dimensions.

You will have to reimplement this to make it work with symbolic shapes.

Comment (RaffEdwardBAH) [inline keras/layers/pooling.py]: IMHO `int_shape` should not be a documented part of the backend then. And begin with `__int_shape` to denote that it should be treated as private. But that is a separate issue. 

How would I make this work for symbolic shapes when `reapeat_elements` didn't work with tensorflow's output from `shape` ? I have no problem making necessary changes when time is available, but I have no intention on debugging large parts of the Keras backend with Theano and Tensorflow to root out behavior inconsistencies.

**Comment by PiranjaF** [comment]
This is awesome. Thanks! I really hope that it'll be integrated into the main repo.

**Comment by PiranjaF** [comment]
I've been trying to use this across some models. This GlobalMaxPooling1D seems to be quite slow. For a large model with maxlen=30 the GlobalMaxPooling1D is 25% slower pr. epoch than using GlobalAveragePooling1D with everything else kept untouched (other layers, data, batchsize etc.). I'm using Theano.

@RaffEdwardBAH , do you have any ideas on how GlobalMaxPooling1D could be optimized?

**Comment by pifelbo** [comment]
For anyone interested: I've found that the above implementation of the GlobalMaxPooling1D layer does not work properly. You can test it using the following code:

from __future__ import print_function, division

import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Masking
from keras.layers import GlobalMaxPooling1D, GlobalAveragePooling1D
from keras.preprocessing.sequence import pad_sequences

add_data_len = 2
maxlen = 3
feats_pr_step = 5

# define model
mask_max = Sequential()
mask_max.add(Masking(input_shape=(maxlen, feats_pr_step)))
mask_max.add(GlobalMaxPooling1D())

mask_mean = Sequential()
mask_mean.add(Masking(input_shape=(maxlen, feats_pr_step)))
mask_mean.add(GlobalAveragePooling1D())

# create test cases
N = 3
comb = add_data_len * feats_pr_step

for dt in [np.int, np.float]:
    test_cases = np.arange(-N/2 * comb, N/2 * comb, dtype=dt).reshape(N, add_data_len, feats_pr_step)

    for op in [[np.max, mask_max], [np.mean, mask_mean]]:
        print(op[0].__name__ + ' ' + str(dt))
        np_func, model = op[0], op[1]
        goal = np_func(test_cases, axis=1)
        test_cases_pad = pad_sequences(test_cases, maxlen=maxlen, dtype=dt,
                                       padding='post',
                                       truncating='post')
        pred = model.predict(test_cases_pad, batch_size=1)
        print(test_cases_pad)
        print("Goal")
        print(goal)
        print("Pred")
        print(pred)
        assert np.all(np.isclose(pred, goal))

**Comment by fchollet** [comment]
Closing outdated PR. If you still care about the content of the PR, please submit a new PR to `master`, updated for the Keras 2.0 API.
